### PR TITLE
Safe js hacks

### PR DESCRIPTION
**Pull request by arjunguha**

### CODE REVIEW

File: js/src/main/scala/JavaScriptCEKLike.scala [hunk 1]
```diff
@@ -337,7 +338,7 @@ trait JavaScriptCEKLike extends slogging.LazyLogging  {
     }
     // This case allows the AST produced by desugaring to insert direct calls to builtin functions.
     case VBuiltinFunction(name) => {
-      if (actuals.forall(_.notFaceted)) {
+      if (actuals.forall(v => v.asInstanceOf[VFacet] == false)) {
```
Comment (rachitnigam) [inline js/src/main/scala/JavaScriptCEKLike.scala]: Should this be `isInstanceOf`?

Comment (arjunguha) [inline js/src/main/scala/JavaScriptCEKLike.scala]: thanks, good catch

File: jvm/src/main/scala/RhinoToQuark.scala [hunk 2]
```diff
@@ -261,8 +261,7 @@ object RhinoToQuark {
       case SForInLoop(x, _, obj, body) => {
         val keys = "#keys"
         val i = "#i"
-        // TODO(arjun): Th efor
```
Comment (rachitnigam) [inline jvm/src/main/scala/RhinoToQuark.scala]: What?

Comment (arjunguha) [inline jvm/src/main/scala/RhinoToQuark.scala]: no idea. deleted the comment.

**Comment by arjunguha** [comment]
This is a great test to fail. Why don't you debug it @rachitnigam?

**Comment by rachitnigam** [comment]
Okay

File: js/src/main/scala/JavaScriptCEKLike.scala [hunk 3]
```diff
@@ -179,31 +177,9 @@ trait JavaScriptCEKLike extends slogging.LazyLogging  {
...
-          store.update(a, VArray(arr, len = x.toInt + 1, obj))
-        }
-        case VConst(CNum(x)) => { //special case where x is much larger than elems.length
-          val arr = MSeq.tabulate(x.toInt + 1){ i => if(i < len) elems(i) else VConst(CUndefined()) }
-          arr.update(x.toInt, v)
-          store.update(a, VArray(arr, len = x.toInt + 1, obj))
-        }
-        case _ => store.update(a, VArray(elems, len, VObj(fields + (anyToStr(key) -> v))))
+      case VArray(elems, obj@VObj(fields)) => key match {
+        case VConst(CNum(x)) if x.toInt == x => elems.update(x.toInt, v)
```
Comment (rachitnigam) [inline js/src/main/scala/JavaScriptCEKLike.scala]: This doesn't seem to be actually updating the value of elems:
```
before update VConst(CNum(1)),VConst(CNum(2)),VConst(CNum(3))
after update VConst(CNum(1)),VConst(CNum(2)),VConst(CNum(3))
```

**Comment by rachitnigam** [comment]
This doesn't seem to be actually updating the value of elems when the index is negative:
```
before update VConst(CNum(1)),VConst(CNum(2)),VConst(CNum(3))
after update VConst(CNum(1)),VConst(CNum(2)),VConst(CNum(3))
```

**Comment by rachitnigam** [comment]
This is what scalaJS has to say about negative indices:
```
If `index` is negative, it is treated as that number of elements starting from the end of the array.
```
which does not seem to be the consistent with JS semantics:
```
> var arr = [1,2,3]
undefined
> arr[-1]
undefined
> arr
[ 1, 2, 3 ]
> arr[-1] = 4
4
> arr[-1]
4
```

**Comment by rachitnigam** [comment]
It does seem to make the tests slower for some reason.

**Comment by arjunguha** [comment]
Wow, that's remarkable. I expected it to go faster. Go figure....

**Comment by rachitnigam** [comment]
...

**Pull request merged**